### PR TITLE
Save last update check

### LINKED ISSUE #41: Make checking for updates optional
Add a property to the configuration file to disable checking for updates.

Also, don't check for updates if the last check was less than 10 min ago, and handle network errors (e.g. due to no connection) gracefully.

I'd like to implement this if that's okay.

### PR DESCRIPTION
**Pull request by Aloso**
Extends the config file to store

- ~if an update is available (bool)~
- when we last checked for updates (unix timestamp)
- if checking for updates should be enabled (bool)

Now, we only check for updates if ~no update is available according to the config file,~ checking for updates isn't disabled and the last check was at least 24 hours ago.

Fixes #41

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -304,32 +302,48 @@ fn main() {
...
-	});
-	let update_check_done_clone = update_check_done;
-	let help_screen_clone = help_screen;
+
+	let update_checker_join_handle = {
+		let updates = &mut config.lock().unwrap().updates;
+		let config = config.clone();
+		let update_available = update_available.clone();
+		let update_check_done = update_check_done.clone();
+
```
Comment (ArturKovacs) [inline src/main.rs:311]: Your version is a lot nicer than what I did initially.

File: src/configuration.rs [hunk 2]
```diff
@@ -12,10 +13,40 @@ pub struct WindowSection {
 	pub win_y: i32,
 }
 
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct UpdateSection {
+	pub has_update: bool,
```
[the remark this answers is not in the document]
Comment (Aloso) [reply src/configuration.rs:18]: My reasoning was that, if we already know that an update is available, we don't need to perform a network request. However, I realized that this won't work if the program is updated and the config file isn't modified, so I'll remove it.

File: src/main.rs (hunk 1 above)
Comment (Aloso) [reply src/main.rs:311]: Thanks!

File: src/configuration.rs [hunk 3]
```diff
@@ -42,6 +73,11 @@ impl Default for Configuration {
 		Configuration {
 			window: WindowSection { dark: false, win_w: 580, win_h: 558, win_x: 64, win_y: 64 },
 			bindings: None,
+			updates: UpdateSection {
+				has_update: false,
+				disable_update_check: false,
```
[the remark this answers is not in the document]
Comment (Aloso) [reply src/configuration.rs:78]: Fixed.

File: src/configuration.rs [hunk 4]
```diff
@@ -33,30 +44,47 @@ impl UpdateSection {
...
 	pub bindings: Option<BTreeMap<String, Vec<String>>>,
 	pub updates: UpdateSection,
 }
 
+#[derive(Deserialize)]
+pub struct IncompleteConfiguration {
+	pub window: Option<WindowSection>,
+	pub bindings: Option<BTreeMap<String, Vec<String>>>,
+	pub updates: Option<UpdateSection>,
+}
```
Comment (ArturKovacs) [inline src/configuration.rs]: I don't mean to be nitpicking but it's quite suspicous that there are two alomst identical structs. I'm wondering why don't we just make the `UpdateSection` optional i.e. `Option<UpdateSection>`. If it's a `None`, the updater code can simply do the default behaviour which is to check for updates and then set the update section to the approproate `Some`. I peeked into #48 just to see if you did it in a different way there but something similar appears there as well.

Comment (Aloso) [reply src/configuration.rs]: The `Option`s are needed for forwards compatibility, if new options are added in the future. I convert it to a struct without `Option`s so we don't need `unwrap` everywhere.

Comment (ArturKovacs) [reply src/configuration.rs]: This seems to me that it comes down to personal preference as this is about having a few extra pattern matches vs having an additionnal struct. I appreciate the time you invested working on this so I'll accept your decision and I'll merge this as it is.

**Pull request merged**